File: happy-ladybugs/happy-ladybugs-1.c
#include <assert.h>
#include <stdint.h>
#include <stdio.h>

int happy_ladybug(const char* b) {
  size_t d[26] = {0};
  size_t has_empty = 0;
  size_t already_happy = 1;
  size_t adj_count = 0;
  size_t left_c = 0;

  for (; *b != 0; ++b) {
    size_t c = *b;

    if (c == '_') {
      has_empty = 1;
    } else {
      d[c - 'A'] += 1;

      if (c == left_c) {
        adj_count += 1;
      }

      if ((adj_count < 2) && (left_c != 0)) {
        already_happy = 0;
      }

      if (c != left_c) {
        adj_count = 1;
      }

      left_c = c;
    }
  }

  if ((adj_count < 2) && (left_c != 0)) {
    already_happy = 0;
  }

  if (already_happy) {
    return 1;
  }

  for (size_t i = 0; i < 26; ++i) {
    if (d[i] == 1) {
      return 0;
    }
  }

  return has_empty;
}

int main() {
  assert(happy_ladybug("_") == 1);
  assert(happy_ladybug("AA") == 1);
  assert(happy_ladybug("A_BBA") == 1);
  assert(happy_ladybug("_ABBA") == 1);
  assert(happy_ladybug("AABB") == 1);
  assert(happy_ladybug("__AA") == 1);
  assert(happy_ladybug("BB__AA") == 1);
  assert(happy_ladybug("BB_AA") == 1);
  assert(happy_ladybug("BBAA_") == 1);
  assert(happy_ladybug("A") == 0);
  assert(happy_ladybug("_A") == 0);
  assert(happy_ladybug("AABBCD") == 0);
  assert(happy_ladybug("ABAB") == 0);
  assert(happy_ladybug("X_") == 0);
  assert(happy_ladybug("X_Y_Z_") == 0);
  assert(happy_ladybug("AABBC") == 0);
}

// def main():

// main()

File: happy-ladybugs/happy-ladybugs-1.py
#!/bin/python3

def happy_ladybug(b):
    if len(b) == 0:
        return True

    left_c = b[0]
    b = b[1:]
    d = {}
    has_empty = left_c == '_'
    already_happy = True
    adj_count = 1

    if left_c != '_':
        d[left_c] = d.setdefault(left_c, 0) + 1

    for c in b:
        if c == '_':
            has_empty = True
            if adj_count < 2:
                already_happy = False
        else:
            d[c] = d.setdefault(c, 0) + 1

            if c == left_c:
                adj_count += 1
            else:
                if adj_count < 2:
                    already_happy = False
                adj_count = 1
            left_c = c

    if adj_count < 2:
        already_happy = False

    all_at_least_two = all([x >= 2 for x in d.values()])

    return already_happy or (has_empty and all_at_least_two)


def main():
    assert happy_ladybug("")
    assert happy_ladybug("A_BBA")
    assert happy_ladybug("_ABBA")
    assert happy_ladybug("AABB")
    assert happy_ladybug("__AA")
    assert happy_ladybug("BB__AA")
    assert happy_ladybug("BB_AA")
    assert happy_ladybug("BBAA_")
    assert not happy_ladybug("ABAB")
    assert not happy_ladybug("X_")
    assert not happy_ladybug("X_Y_Z_")
    assert not happy_ladybug("AABBC")


main()
